SimBrief Planning System Updates

FatihKoz opened the pull request:
**1. flight / table ;**

Added aircraft selection for bidded flight
SimBrief button is not visible until aircraft selection
Aircraft_id is passed to simbrief_form with flight_id as qstring

**2. flight / simbrief_form ;**

Uses the aircraft_id to get all necessary details to calculate random load according to load factor and variances. (Min and Max protections applied, load factor will never go above 100 and below 1)
All possible values are sent to simbrief for more precise and customized flight plan generation
Gets current utc time and adds 45mins to it for ETD and Date generation
Different pax weights are used for Charter and Scheduled flights (also sent to SimBrief)
Calculated random load figures are sent as Dispatch Remark too
Automatic SID/STAR type selection added to planning options (RNAV or Conventional)
Step Climb logic applied (as per SimBrief documentation, if Step Climbs are enabled no flight level info is needed)
Gets user weight settings (KG/LBS) and uses it on both calculations and SimBrief Briefing Options section

**3. flight / simbrief_briefing ;**

EET display corrected, it was using scheduled time enroute but label/intention was to display estimated time (compared to my previous and merged PR)

**4. flight / search ;**

Sorted the airlines and subfeets by their name (id sort was not looking good on the dropdown, generating an unsorted list effect)

I think that's all ...

nabeelio commented on resources/views/layouts/default/flights/simbrief_form.blade.php:
```diff
@@ -1,269 +1,426 @@
+
+@php 
+	$SelectedAircraft = app('request')->input('aircraft_id') ;
+	$AircraftDetails = DB::table('aircraft')->select('registration', 'icao', 'iata', 'subfleet_id')->where('id', $SelectedAircraft)->get() ;
```
Can you move this so it's passed in by the controller? And query with the `Aircraft` model instead of using `DB` directly

FatihKoz replied:
Need help with this ... ( not stopped trying to manage myself of course )

FatihKoz replied:
Managed to move the db selection to controller, works fine on my test setup.

nabeelio commented on resources/views/layouts/default/flights/simbrief_form.blade.php:
```diff
@@ -1,269 +1,426 @@
+	if($FKMin < 1) { $FKMin = 1 ; }	
+	if($FKMax > 100) { $FKMax = 100 ; }
+	$FKRandomLoad = rand($FKMin, $FKMax);
+	$GetMaxCap = DB::table('subfleet_fare')->where('subfleet_id', $SubFLid)->sum('capacity') ;
```
This logic should also move into the controller and get passed in

FatihKoz replied:
Removed the GetMaxCap completely

nabeelio commented on resources/views/layouts/default/flights/simbrief_form.blade.php:
```diff
@@ -1,269 +1,426 @@
+					<br>
+						<div class="row">
+							@php
+							$FlightType = 'SimBrief Standart';
```
Spelling of Standard?

FatihKoz replied:
Fixed

nabeelio commented on resources/views/layouts/default/flights/simbrief_form.blade.php:
```diff
@@ -1,269 +1,426 @@
+						<div class="row">
+							@php
+							$FlightType = 'SimBrief Standart';
+							if($flight->flight_type == 'J') { $FlightType = 'Schedule All Adult Pax' ;}
```
Use the ENUMs for these (look for `FlightType` being used elsewhere, it's better to use those constants than hardcoding the 'J', 'G', etc

FatihKoz replied:
Please check the new commit.

nabeelio commented on resources/views/layouts/default/flights/table.blade.php:
```diff
@@ -1,91 +1,113 @@
+					<select id="AircraftSelection" class="form-control select2" onchange="OnChangeACSelection()">
+						<option value="ZZZZZ">Please Select An Aircraft</option>
+						@if($flight->subfleets)
+							@foreach($flight->subfleets as $SubFleets)
```
The naming convention for these is all lowercase, same for other variable names, so `$subfleet` for this (since it's a single one we're looping through). StyleCI doesn't check these files but the naming should match across the board

FatihKoz replied:
Fixed as requested like others

nabeelio commented on resources/views/layouts/default/flights/table.blade.php:
```diff
@@ -1,91 +1,113 @@
+				}else{
+					document.getElementById('SbFormLink').style.visibility = 'visible';
+				}
+			var $OldLink = document.getElementById("SbFormLink").href ;
```
Use `const` instead of `var`

FatihKoz replied:
fixed the javascript logic and variable names in table.blade.php , sent it as a commit to this pr.

nabeelio commented on resources/views/layouts/default/flights/simbrief_form.blade.php:
```diff
@@ -1,269 +1,426 @@
+					
+					<div class="form-container-body">
+					{{-- Get All Subfleets from flight and generate random load for each fare type of selected SubFleet --}}
+					@foreach($flight->subfleets as $SUB)
```
Use `$subfleet`. Trying to keep these the same conventions/naming across the codebase

FatihKoz replied:
Fixed.

nabeelio commented on resources/views/layouts/default/flights/simbrief_form.blade.php:
```diff
@@ -1,269 +1,426 @@
+						{{-- Generate Load Figures --}}
+						<div class="row">
+						@php $LoadArray = [] ; @endphp
+						@foreach($SUB->fares as $SUBfares)
```
Instead of `SUBfares` use `fare`

FatihKoz replied:
Done

nabeelio commented:
Looks great, just some cleanup of variable names across the board to keep it the same/aligned across the codebase, and moving some stuff into the controllers so there aren't DB lookups in templates - that prevents the templates from being cached, but also keeps the layers separated

FatihKoz commented:
Thanks for the detailed replies @nabeelio , honestly I tried to update the controller for the db reads I needed but simply can not succeeded. I am not good at laravel and its logic , thus I used direct reads sorry.

Will fix the others though

FatihKoz commented:
Fixed the variable name issues and removed the GetMaxCap direct db read completely ... Will look into the Enums for flight type , but really need advice/examples/helps on the controller change to be honest.

We only need one change required in the controller to get the aircraft details, will work on it though but not sure about the result :)

I know laravel does not want pure php usage in the view side and wants all that in the controller as much as possible but this is what i can do right now with my limited laravel knowledge.

nabeelio commented:
Is this OK to merge? Or do you still have stuff to add?

nabeelio commented:
I did some cleanup on this (tabs to spaces, moving some logic to the controller). Leaving some comments inline about things I'm not sure of that need to be cleaned up.

nabeelio commented on resources/views/layouts/default/flights/simbrief_form.blade.php:
```diff
@@ -1,269 +1,411 @@
+      if($acdetails->icao == 'A21N') { $simbrieftype = 'A321' ; }
+      if($acdetails->icao == 'B77L') { $simbrieftype = 'B77F' ; }
+      if($acdetails->icao == 'B773') { $simbrieftype = 'B77W' ; }
+      if($acdetails->icao == 'E35L') { $simbrieftype = 'E135' ; }
```
What about additional types? This will only set the `$simbrieftype` based on the last aircraft in the list

FatihKoz replied:
$simbrieftype = $acdetails->icao ;

It is already being assigned before the if's (just 2 lines above the first if) , these are the changes where we make the icao type compatible with simbrief ... They are the ones I found where simbrief uses different type codes for an icao code

nabeelio commented on resources/views/layouts/default/flights/simbrief_form.blade.php:
```diff
@@ -1,269 +1,411 @@
+                </div>
+                <br>
+                <div class="form-container-body">
+                  <h6><i class="fas fa-info-circle"></i>&nbsp;@lang('stisla.briefingoptions')</h6>
```
Any skin-specific stuff shouldn't be here

FatihKoz replied:
Sorry about it, missed it while checking

nabeelio commented on resources/views/layouts/default/flights/simbrief_form.blade.php:
```diff
@@ -1,269 +1,411 @@
+            @if($totalgenload > 0)
+              @php
+                $loaddisttxt =  "Load Distribution ";
+                $loaddist = implode(' ', array_map(
```
What is this doing? What exactly is being mapped?

FatihKoz replied:
It is mapping the loadarray array , which is created above before generating loads for each fare type assigned to the aircraft's subfleet , it holds the farecode and the generated random load for that fare ... Then used here to create the load distribution message sent to simbrief as manualrmk

FatihKoz replied:
I think we have no more issues on this, if you do not want it in the dev we can remove it too.

Not the loadarray of course it is needed for the calculations itself but we can remove the manualrmk section too as we did for the info text.

Personally I like the idea keeping it, maybe some acars developer reads the dispatch remarks from the simbrief.xml and uses it somewhere ;)

nabeelio commented on resources/views/layouts/default/flights/simbrief_form.blade.php:
```diff
@@ -1,269 +1,411 @@
+                      $flightype = 'SimBrief Standard';
+                      if($flight->flight_type == 'C') { $flightype = 'Charter All Adult Pax' ;}
+                      if($flight->flight_type == 'J' || $flight->flight_type == 'G') { $flightype = 'Schedule All Adult Pax' ;}
+                      if($flight->flight_type == 'F' || $flight->flight_type == 'A' || $flight->flight_type == 'H') { $flightype = 'Only Cargo' ;}
```
This should all use the language strings - so below should read `{{ FlightType::label($flight->flight_type) }}`

FatihKoz replied:
Why ? I am not using the flight type here 

This is where I build an info text according to the flight type , I agree on using language files but then we need to add 'Schedule All Adult Pax' this to the language pack.

Or this can be completely removed , it is just an info text displayed to give the end user info about which weights will be used in his simbrief flight plan

nabeelio replied:
Is "Schedule All Adult Pax" industry language? Just asking because I'm not familiar with it. If you want it to be here then I think it makes sense to then add it to the language files. There is language there now... ex, `lang/en/flights.php`, "Passenger - Scheduled" - could that be used? "Passenger - Scheduled All Adult Pax", for example?

FatihKoz replied:
Well, the actually industry specific name is "All flights except holiday charters" but it is like the VMC/IMC thing, on daily basis people use "Scheduled" or "Charter" mostly , and it can be varied as "All Adult" or "Male/Female/Children" ... 

![mb](https://user-images.githubusercontent.com/74361521/102833485-4bf67b00-4402-11eb-9d7d-1bc792a06ffb.png)

FatihKoz replied:
Yeah , it is just an info text after all , may be removed I can handle it on my side :) No need to go this industry specific

FatihKoz replied:
Removed the info text completely , will handle it on my site personally

nabeelio commented on resources/views/layouts/default/flights/table.blade.php:
```diff
@@ -69,23 +62,64 @@
+                    @if($flight->subfleets->isNotEmpty())
+                      @php $subfleets = $flight->subfleets; @endphp
+                    @else
+                      @php $userm = Auth::user(); $userSvc = app(App\Services\UserService::class); $subfleets = $userSvc->getAllowableSubfleets($userm); @endphp
```
Which page is this showing up on? This needs to be moved to the controller also like I did before. Is it just the flight index?

nabeelio replied:
Oof this is within the context of a flight. I think this should be on the briefing page only, not here.

FatihKoz replied:
This is the table.blade in flights , but we need to have a solution here too if we want to select the aircraft before reaching the simbrief_form.blade

And this the solution where va owner does not assign any fleets to his flights , then we need to get the list from the user

nabeelio replied:
Can this just be moved/handled only in the simbrief form? This is really heavy logic to put in here - for every flight, it's 1 call to get subfleets, then the fares for each subfleet, so a minimum of 2 calls - with 20 flights that's 40 separate calls, which aren't cached. And this also prevents the template from being cached.

FatihKoz replied:
As I said , we need a solution to get the aircraft list before the simbrief_form.blade and this is displayed for bidded flights if the va owner does not assigns any subfleets to his flights ... 

If you move this to the simbrief_form , then we need to have a dropdown there to select the aircraft (which was my first solution and did not worked well to be honest)

If we can improve the flights controller and return the flights collection with the aircrafts according to what user can fly in the first place and handling situations where the va owner does not assign any subfleets to his flights then this will not be needed at all

nabeelio replied:
I can probably do that but I think moving it to the briefing page is the right place for it, not in this main flight listing. Can you remove this from here and move it to the simbrief form? The code should just be able to be moved. And then I will fix the controller.

FatihKoz replied:
Result where admin assigns fleet/fleets to a flight (we get the flight->subfleets lists)

![ac1](https://user-images.githubusercontent.com/74361521/102832878-b5758a00-4400-11eb-8f2b-dee1a540d4cf.png)

Where not (we get users aircraft list)

![ac2](https://user-images.githubusercontent.com/74361521/102832950-d1792b80-4400-11eb-8443-7b006fa76d49.png)

If we can do this in the controller it would be great of course

nabeelio replied:
Yea in the default skin it would be too much there. So I think just move that to the simbrief form, as-is, and I'll fix in the controller. I moved other stuff into the controller so if you can pull this PR down and make sure it's all still working, that'd be great

FatihKoz replied:
So you do not want to have aircraft selection in here at all , and all be done in simbrief_form ? did I understand correctly

nabeelio replied:
Correct, yes in the simbrief_form.

FatihKoz replied:
Well , we are going back to where we started then ...

Anyway, i will look into it , if no aircraft_id is passed then I can render the selection dropdown and when user selects his aicraft and with a kind page reload he will see the final simbrief_form with all data mathching the flight and the selected aircraft.

Do not want to use javascript for aircraft selection (as the first thing I did in the beginning) , anywill work on it.

So are we closing this whole PR completely ?

nabeelio replied:
No, this PR is fine. What's wrong with moving the aircraft dropdown code that's in the flight selection into the simbrief form template? It should be able to just move over. I can look into doing it if you can't get it

FatihKoz replied:
It was where we started, we needed the aicraft selection before all the calculations on simbrief_form.blade :) Anyway I need to work a little bit on it ... But I think I can handle it with a reload on the simbrief_form.blade upon aircraft selection if it fine for you too

nabeelio replied:
Ok don't do anything yet them - let me think of a way around this.

nabeelio replied:
If there was a selection page before the form, would that help? Would there be other things to add to that page, aside from selecting the aircraft, that would be helpful?

FatihKoz replied:
The load calculations and the form fields required for simbrief plan generation relies on the aircraft selection (and of course subfleet it is a member of) 

But if a reload is ok, we can check it in the controller and if an aircraft is is not passed we can open the selection dropdown , upon selection page will reload itself with the id and goes on

nabeelio replied:
A reload is brittle. Lemme try adding an aircraft selection page before the simbrief page. I think that will be better, what do you think? So click "generate simbrief", a page to select the aircraft, and then the simbrief form? Lemme also look at which load calculations, maybe it's ok to do in JS on selection. But that is probably messy, unless it can load the extra data from the API. I do that on the PIREP form itself - if you look at the aircraft selection on the PIREP form, I load the fare information dynamically that way.

FatihKoz replied:
I tried to this all in the simbrief_form with a dropdown and JS , it did not worked well ... Not impossible of course but I ended up implementing this solution 'cause did not like the final result and capabilities of enduser calculations. I wanted and needed all the calculations done on the server side to be honest.

A page before simbrief_form which passes the flight_id and the aircraft_id would work of course but I think we can handle it in the simbrief_form too , in the end user will see either 2 different pages or a page loaded second time. Not much a difference for the end user

FatihKoz replied:
Currently we are checking the flight_id in the controler and I added the aircraft_id check to it too ... So if there is no aicraft_id passed I will open the dropdown , send the selection back to the page with &aircraft_id= as I did in the dropdown at table.blade 

If no aircraft_id is present , current controller was redirecting  the user to the bids page , I will remove that part and put a flash alert message and show the dropdown (rest of the page will not be rendered of course)

FatihKoz replied:
How about a solution working like this ...

User reaches the simbrief_form without an aircraft selection
![sb1](https://user-images.githubusercontent.com/74361521/102839083-c4fccf00-4410-11eb-9f92-e3c38437d8fc.png)

Selects the aircraft here
![sb2](https://user-images.githubusercontent.com/74361521/102839094-c8905600-4410-11eb-943b-34d5073daffb.png)

Reaches the simbrief_form again with his flight and aircraft_id's returned
![sb3](https://user-images.githubusercontent.com/74361521/102839102-cc23dd00-4410-11eb-8105-c3333ee16d42.png)

With this, we can completely remove the aircraft selection dropdown from the flights.table_blade

FatihKoz commented:
So i think this is better than having a seperate page in between flights.table and flights.simbrief_form.

With the controller updated to overcome the problems we had while selecting the aircraft and subfleet fares, this will be ok for me to be merged with the dev. I have no additions for this PR.

FatihKoz commented:
I think we should close this pr and open a new one, which will be compatible with the latest dev and much easy to follow up ...

1/ flight.table : not using it anymore ( selection moved the simbrief_form )
2/ flight.simbrief_briefing : The change was not related to the main update and outdated now ( generate new briefing pr )
3/ flight.search : This was also not related to the main update and outdated too ( flight type search pr )

What we need is the simbreif.controller and simbrief_briefing.blade files and controller must be updated according to latest dev build.

nabeelio commented:
Sounds good to me
